tests/main/views/test_audits.py

```diff
@@ -50,8 +54,204 @@ def add_audit_events_with_db_object(self):
+                    izip(repeat((Supplier, supplier_ids,)), supplier_audit_event_params),
+                    ):
+                ae = AuditEvent(audit_type, "henry.flower@example.com", {}, ref_model(id=ref_model_ids[obj_id]))
+                ae.created_at = created_at
+                ae.acknowledged_at = acknowledged_at
+                ae.acknowledged = bool(acknowledged_at)
+                db.session.add(ae)
+                audit_events.append(ae)
+
+            db.session.commit()
+            # make a note of the ids that were given to these events, or rather the order they were generated
+            audit_event_id_lookup = {ae.id: i for i, ae in enumerate(audit_events)}
+            assert AuditEvent.query.count() == len(service_audit_event_params)+len(supplier_audit_event_params)
+
+            return audit_event_id_lookup
+
 
 class TestAuditEvents(BaseTestAuditEvents):
+    @pytest.mark.parametrize(
+        "service_audit_event_params,supplier_audit_event_params,req_params,expected_resp_events",
+        # where we refer to "id"s in the expected_response_params, because we can't be too sure about the *actual* ids
+        # given to objects, we're using 0-based notional "ids" mased on the order the audit events were inserted into
+        # the db. service_audit_event_params events are inserted in the order given, followed by the
+        # supplier_audit_event_params events. so if we had 5 service_audit_event_params and 2
+        # supplier_audit_event_params, "5" would refer to the audit event created by the first-listed
+        # supplier_audit_event_params.
+        # similarly, where supplier and service "id"s are referred to, the "id"s we're referring to are normalized
+        # pseudo-ids from 0-4 inclusive
+        chain.from_iterable(
+            ((serv_aeps, supp_aeps, req_params, expected_resp_events) for req_params, expected_resp_events in req_cases)
+            for serv_aeps, supp_aeps, req_cases in (
+                (
+                    (  # service_audit_event_params
+                        (0, AuditTypes.update_service, datetime(2010, 6, 6), None,),
+                        (0, AuditTypes.update_service, datetime(2010, 6, 7), None,),
+                        (4, AuditTypes.update_service, datetime(2010, 6, 2), None,),
+                    ),
+                    (  # supplier_audit_event_params
+                        (0, AuditTypes.supplier_update, datetime(2010, 6, 6), None,),
+                    ),
+                    (  # and now a series of req_cases - pairs of (req_params, expected_resp_events) to test against
+                       # the above db scenario. these get flattened out into concrete test scenarios by
+                       # chain.from_iterable above before they reach pytest's parametrization
+                        (
+                            {"earliest_for_each_object": "true", "latest_first": "true"},
+                            (3, 0, 2,),
```

Comment:
This is pretty minor, but I think using tuples and lists semantically here (as opposed to always using tuples) would help visually parse this structure. Eg `(serv_aeps, supp_aeps, req_params, expected_resp_events)` should be a tuple, since each positional element has a different meaning, but `[3, 0, 2]` is a list of ordered ids.

I'm guessing you're doing this to avoid accidental mutation, but the code that's using these values in the `assert` is pretty minimal, and we're only using them once so it doesn't feel like a risk at all.